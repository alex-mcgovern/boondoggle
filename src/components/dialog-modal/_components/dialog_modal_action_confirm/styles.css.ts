import { style } from "@vanilla-extract/css";
import { utilCss } from "../../../../index.css";

export const confirmTextStyle = style([
	utilCss({
		display: "inline-block",
		marginBottom: "space_2",
	}),
	{
		userSelect: "none",
	},
]);
