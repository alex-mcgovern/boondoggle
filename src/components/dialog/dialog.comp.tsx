import { Slot } from "@radix-ui/react-slot";
import clsx from "clsx";
import { createRef, forwardRef, useCallback } from "react";

import { useClickOutside } from "../../hooks/use_click_outside";
import { useForwardRef } from "../../hooks/use_forward_ref";
import { useOpenDialogWithKeyboard } from "../../hooks/use_open_dialog_with_keyboard";
import { a11yFocus } from "../../styles/common/a11y.css";
import { animateAppear } from "../../styles/common/animations.css";
import { Box } from "../box";
import { getDialogStyles } from "./dialog.styles.css";

import type { BoxProps } from "../box";
import type { DialogPlacementEnum } from "./dialog.styles.css";
import type { ReactNode } from "react";

export type DialogProps = BoxProps & {
  children?: ReactNode;
  className?: string;
  inert?: boolean;
  isOpen?: boolean;
  onIsOpenChange?: (isOpen?: boolean) => void;
  placement?: DialogPlacementEnum;
  preventOpenOnKeydown?: boolean;
  triggerNode?: ReactNode;
  wrapperProps?: BoxProps;
};

export const Dialog = forwardRef<HTMLDialogElement, DialogProps>(
  (
    {
      children,
      className: userClassName,
      isOpen: controlledIsOpen,
      onIsOpenChange,
      preventOpenOnKeydown = false,
      triggerNode,
      placement,
      wrapperProps,
      ...rest
    }: DialogProps,
    ref
  ) => {
    const dialogRef = useForwardRef<HTMLDialogElement>(ref);

    const triggerRef = createRef<HTMLElement>();

    /**
     * Callback for when the input is clicked or focused.
     */
    const toggleIsOpen = useCallback(() => {
      if (!dialogRef.current) {
        return undefined;
      }

      return (
        dialogRef.current?.open
          ? dialogRef.current?.close()
          : dialogRef.current?.show(),
        onIsOpenChange?.(dialogRef.current?.open)
      );
    }, [dialogRef, onIsOpenChange]);

    /**
     * When input is focused and user presses enter, open the date picker.
     */
    useOpenDialogWithKeyboard<HTMLElement>({
      callback: toggleIsOpen,
      dialogRef,
      preventOpenOnKeydown,
      triggerRef,
    });

    /**
     * Handle click outside dialog
     */
    useClickOutside<HTMLDialogElement, HTMLElement>({
      callback: () => {
        return dialogRef.current?.close();
      },
      contentRef: dialogRef,
      triggerRef,
    });

    return (
      <Box position="relative" {...wrapperProps}>
        <Slot onClick={toggleIsOpen} ref={triggerRef}>
          {triggerNode}
        </Slot>
        <Box
          {...rest}
          as="dialog"
          open={controlledIsOpen}
          ref={dialogRef}
          className={clsx(
            animateAppear,
            userClassName,
            a11yFocus,
            getDialogStyles({ placement })
          )}
        >
          {children}
        </Box>
      </Box>
    );
  }
);
