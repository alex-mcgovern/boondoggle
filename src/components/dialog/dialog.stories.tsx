import React from "react";

import { Dialog as StoryComponent } from ".";
import { WithDarkLightMode } from "../../../.storybook/with_dark_light_mode";
import { LOREM } from "../../../mocks/LOREM.mock";
import { Button } from "../button";

import type { DialogProps as StoryComponentProps } from ".";
import type { StoryFn, StoryObj } from "@storybook/react";

export default {
  title: "React components/Dialog",
  component: StoryComponent,
};

const Template: StoryFn<StoryComponentProps> = ({
  ...rest
}: StoryComponentProps) => {
  return <StoryComponent {...rest} />;
};

export const Default: StoryObj<StoryComponentProps> = {
  render: Template,
  name: "default",
  parameters: {
    layout: "fullscreen",
  },
  decorators: [WithDarkLightMode],
  args: {
    triggerNode: <Button name="Dialog button">Open dialog</Button>,
    children: LOREM.text_xxl,
    title: LOREM.text_xxs,
  },
};
