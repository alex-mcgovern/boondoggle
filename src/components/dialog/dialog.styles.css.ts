import { keyframes, style } from "@vanilla-extract/css";

import { vars } from "../../styles/theme.css";
import { createAccessibleTransition } from "../../styles/utils/create_accessible_transition";
import { getSprinkles } from "../../styles/utils/get_sprinkles.css";

const overlayShow = keyframes({
  "0%": { opacity: 0 },
  "100%": { opacity: 1 },
});

const contentShow = keyframes({
  "0%": { opacity: 0, transform: "translate(-50%, -48%) scale(.96)" },
  "100%": { opacity: 1, transform: "translate(-50%, -50%) scale(1)" },
});

export const dialogOverlay = style([
  {
    background: `#00000080`,
    position: "fixed",
    inset: 0,
    zIndex: 99999,
  },
  createAccessibleTransition({
    animation: `${overlayShow} ${vars.transitionDuration.medium} cubic-bezier(0.16, 1, 0.3, 1) forwards`,
  }),
]);

export const dialogContent = style([
  getSprinkles({
    display: "block",
    isolation: "isolate",
    overflow: "hidden",
    textDecoration: "none",
    background: "tint_default",
    border: "border_default",
    borderRadius: "md",
    boxShadow: "lg",
    width: {
      mobile: "gridSpan10",
      tablet: "gridSpan8",
      desktop: "gridSpan8",
    },
    maxWidth: "gridSpan10",
  }),
  {
    zIndex: 99999,
    position: "fixed",
    top: "50%",
    left: "50%",
    transform: "translate(-50%, -50%)",
    selectors: {
      "&:focus": { outline: "none" },
    },
  },
  createAccessibleTransition({
    animation: `${contentShow} ${vars.transitionDuration.long} ease forwards`,
  }),
]);

export const dialogInner = style([
  getSprinkles({
    maxHeight: "90vh",
    padding: "spacing2",
  }),
  {
    overflowY: "auto",
  },
]);
