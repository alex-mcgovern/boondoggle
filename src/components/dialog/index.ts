export { Dialog } from "./dialog.component";
export type { DialogProps } from "./dialog.component";
