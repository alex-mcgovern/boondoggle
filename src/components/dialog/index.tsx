import * as RadixSlot from "@radix-ui/react-slot";
import clsx from "clsx";
import { createRef, forwardRef, useCallback } from "react";

import { useClickOutside } from "../../hooks/use-click-outside";
import { useOpenDialogWithKeyboard } from "../../hooks/use-open-dialog-with-keyboard";
import { useForwardRef } from "../../hooks/use-forward-ref";
import { Box } from "../box";
import { getDialogStyles } from "./styles.css";

import type { ReactNode } from "react";
import type { BoxProps } from "../box";
import type { DialogPlacementEnum } from "./styles.css";

export type DialogProps = BoxProps & {
	/**
	 * Children to render inside the dialog component.
	 */
	children?: ReactNode;

	/**
	 * Optional CSS class name applied to the outer HTML dialog element.
	 */
	className?: string;

	/**
	 * Whether the dialog is "inert" (can't be focused, can't be clicked.)
	 */
	inert?: boolean;

	/**
	 * Controls the dialog's open state, making it a controlled element. Use alongside `onIsOpenChange`.
	 */
	isOpen?: boolean;

	/**
	 * Callback when the dialog's open state is changed. Use alongside `isOpen`.
	 */
	onIsOpenChange?: (isOpen?: boolean) => void;

	/**
	 * Controls the interaction mode for the dialog trigger node.
	 */
	openOn?: "click" | "hover";

	/**
	 * Controls the dialog's placement relative to the trigger node.
	 */
	placement?: DialogPlacementEnum;

	/**
	 * Whether to open the dialog on an enter keypress while the trigger node is focused.
	 */
	preventOpenOnKeydown?: boolean;

	/**
	 * The react node to act as the trigger for the dialog.
	 */
	triggerNode?: ReactNode;

	/**
	 * `BoxProps` to pass to the outer element that wraps the dialog and it's trigger.
	 */
	wrapperProps?: BoxProps;
};

export const Dialog = forwardRef<HTMLDialogElement, DialogProps>(
	(
		{
			children,
			className: userClassName,
			isOpen: controlledIsOpen,
			onIsOpenChange,
			openOn = "click",
			placement,
			preventOpenOnKeydown = false,
			triggerNode,
			wrapperProps,
			...rest
		}: DialogProps,
		ref,
	) => {
		const dialogRef = useForwardRef<HTMLDialogElement>(ref);

		const triggerRef = createRef<HTMLElement>();

		const toggleIsOpen = useCallback(() => {
			if (!dialogRef.current) {
				return;
			}

			if (dialogRef.current?.open) {
				dialogRef.current?.close();
			}

			if (!dialogRef.current?.open) {
				dialogRef.current?.show();
			}

			onIsOpenChange?.(dialogRef.current?.open);
		}, [dialogRef, onIsOpenChange]);

		useOpenDialogWithKeyboard<HTMLElement>({
			callback: toggleIsOpen,
			dialogRef,
			preventOpenOnKeydown,
			triggerRef,
		});

		useClickOutside<HTMLDialogElement, HTMLElement>({
			callback: () => {
				return dialogRef.current?.close();
			},
			contentRef: dialogRef,
			triggerRef,
		});

		return (
			<Box position="relative" {...wrapperProps}>
				<Box background="transparent">
					<RadixSlot.Slot
						onClick={toggleIsOpen}
						onMouseEnter={
							openOn === "hover"
								? () => {
										return dialogRef.current?.show();
								  }
								: undefined
						}
						onMouseLeave={
							openOn === "hover"
								? () => {
										return dialogRef.current?.close();
								  }
								: undefined
						}
						onPointerEnter={
							openOn === "hover"
								? () => {
										return dialogRef.current?.show();
								  }
								: undefined
						}
						onPointerLeave={
							openOn === "hover"
								? () => {
										return dialogRef.current?.show();
								  }
								: undefined
						}
						onPointerOver={
							openOn === "hover"
								? () => {
										return dialogRef.current?.show();
								  }
								: undefined
						}
						ref={triggerRef}
					>
						{triggerNode}
					</RadixSlot.Slot>
				</Box>
				<Box
					{...rest}
					as="dialog"
					className={clsx(
						userClassName,
						getDialogStyles({ placement }),
					)}
					open={controlledIsOpen}
					ref={dialogRef}
				>
					<Box>{children}</Box>
				</Box>
			</Box>
		);
	},
);
