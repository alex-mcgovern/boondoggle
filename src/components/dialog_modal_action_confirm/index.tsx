import { forwardRef, useState } from "react";

import { Box } from "../box";
import { Button } from "../button";
import { Input } from "../input";
import { confirmTextStyle } from "./styles.css";

import type { ColorOverlay } from "../../styles/color_palette.css";
import type { ButtonProps } from "../button";

export type DialogModalActionConfirmProps = {
  buttonProps?: Omit<ButtonProps, "onClick">;
  buttonText: string;
  colorOverlay?: ColorOverlay;
  confirmText: string;
  onClick?: ButtonProps["onClick"];
  promptPrefix: string;
  promptSuffix: string;
};

export const DialogModalActionConfirm = forwardRef<
  HTMLDivElement,
  DialogModalActionConfirmProps
>(
  (
    {
      buttonProps,
      buttonText,
      colorOverlay,
      confirmText,
      onClick,
      promptPrefix,
      promptSuffix,
    }: DialogModalActionConfirmProps,
    ref
  ) => {
    const [userConfirmText, setUserConfirmText] = useState("");

    return (
      <Box ref={ref}>
        <Box className={confirmTextStyle}>
          <Box as="span">{promptPrefix}</Box>{" "}
          <Box as="span" fontWeight="bold">
            {confirmText}
          </Box>{" "}
          <Box as="span">{promptSuffix}</Box>
        </Box>

        <Input
          colorOverlay={colorOverlay}
          marginBottom="spacing_2"
          name="dialog_confirm_text"
          onChange={(e) => {
            return setUserConfirmText(e.target.value);
          }}
          placeholder=""
          value={userConfirmText}
        />

        <Button
          appearance="primary"
          colorOverlay={colorOverlay}
          disabled={userConfirmText !== confirmText}
          name="primary_action"
          onClick={onClick}
          width="100%"
          {...buttonProps}
        >
          {buttonText}
        </Button>
      </Box>
    );
  }
);
