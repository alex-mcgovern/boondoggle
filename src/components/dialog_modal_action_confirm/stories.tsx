import { DialogModalActionConfirm as StoryComp } from ".";

import type { Meta, StoryObj } from "@storybook/react";

const meta = {
    args: {
        buttonText: "Freeze account",
        confirmText: "freeze account",
        onClick: () => {
            alert("Confirmed");
        },
        promptPrefix: "Please type",
        promptSuffix: "to continue",
    },
    component: StoryComp,
    title: "Components/DialogModalActionConfirm",
} satisfies Meta<typeof StoryComp>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {};

export const WithColorOverlay: Story = {
    args: {
        colorOverlay: "red",
    },
};
