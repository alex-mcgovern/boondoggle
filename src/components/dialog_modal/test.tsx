/**
 * @jest-environment jsdom
 */
import { render, waitFor } from "@testing-library/react";
import { createRef } from "react";

import { DialogModal } from ".";
import "../../../test/mocked_dependencies/dialog.mock";
import { DialogModalActionConfirm } from "../dialog_modal_action_confirm";

import type { DialogModalProps } from ".";

const renderComponent = async (props: DialogModalProps) => {
    const dialogRef = createRef<HTMLDialogElement>();

    const openDialog = () => {
        dialogRef.current?.showModal();
    };

    return waitFor(() =>
        {return render(
            <>
                <button
                    onClick={openDialog}
                    type="button"
                >
                    Open
                </button>
                <DialogModal
                    ref={dialogRef}
                    {...props}
                />
            </>
        )}
    );
};

describe("<DialogModal />", () => {
    describe("default", () => {
        describe("on initial render", () => {
            test("should render without throwing", async () => {
                const { container } = await renderComponent({
                    children: "hello",

                    title: "hello",
                    width: "lg",
                });

                expect(container).not.toBeNull();
            });
        });
    });
});

describe("<DialogModal />", () => {
    describe("default", () => {
        describe("on initial render", () => {
            test("should render without throwing", async () => {
                const { container } = await renderComponent({
                    actions: (
                        <DialogModalActionConfirm
                            buttonText="Remove team member"
                            colorOverlay="red"
                            confirmText="remove team member"
                            onClick={() => {
                                alert("clicked");
                            }}
                            promptPrefix="Please type"
                            promptSuffix="To continue"
                        />
                    ),
                    children: "hello",

                    title: "hello",
                    width: "lg",
                });

                expect(container).not.toBeNull();
            });
        });
    });
});
