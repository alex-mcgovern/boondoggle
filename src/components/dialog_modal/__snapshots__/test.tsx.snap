// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DialogModal /> default on initial render should match snapshot 1`] = `
<div>
  <button
    type="button"
  >
    Open
  </button>
  <div
    class="get_sprinkles_position_relative_mobile__eiij7bgb"
  >
    <dialog
      class="styles_dialogModalOuterStyle__xbhn3g1 animations_animateFadeIn__apgy12 styles__xbhn3g0 get_sprinkles_background_background__eiij7ba get_sprinkles_borderRadius_lg__eiij7bx get_sprinkles_padding_none_mobile__eiij7bbt"
    >
      <div
        class="styles_getDialogModalInnerStyle__bryhpl3 styles__bryhpl2 get_sprinkles_display_flex_mobile__eiij7b8q get_sprinkles_flexDirection_column_mobile__eiij7b9h get_sprinkles_maxWidth_100vw__eiij7b5y get_sprinkles_overflow_hidden__eiij7b86 get_sprinkles_padding_none_mobile__eiij7bbt get_sprinkles_width_100vw_mobile__eiij7bgq styles_getDialogModalInnerStyle_width_sm__bryhpl5 styles_variantWidth_sm__bryhpl1"
      >
        <div
          class="styles_dialogHeaderStyle__7f53ve1 styles__7f53ve0 get_sprinkles_alignItems_center__eiij7b2 get_sprinkles_borderBottom_border_default__eiij7bt get_sprinkles_display_flex_mobile__eiij7b8q get_sprinkles_justifyContent_space-between__eiij7b3c get_sprinkles_maxWidth_100%__eiij7b5x get_sprinkles_padding_spacing_2_mobile__eiij7bc8"
        >
          <h3
            class="styles_dialogTitleStyle__7f53ve3 styles__7f53ve2 get_sprinkles_marginTop_none__eiij7b50 get_sprinkles_marginBottom_none__eiij7b43 get_sprinkles_fontWeight_semibold__eiij7b2i"
          >
            hello
          </h3>
          <button
            aria-label="Close"
            class="styles_getButtonStyles__1yqzeadr a11y_a11yFocus__r8he5t2 a11y_a11yDisabled__r8he5t1 styles__1yqzeadh get_sprinkles_alignItems_center__eiij7b2 get_sprinkles_borderRadius_sm__eiij7b10 get_sprinkles_display_inline-flex_mobile__eiij7b92 get_sprinkles_gap_spacing_1_mobile__eiij7ba2 get_sprinkles_textDecoration_none__eiij7b8d styles_getButtonStyles_appearance_ghost__1yqzeads styles_variantAppearance_ghost__1yqzeadc styles__1yqzead7 get_sprinkles_fontWeight_normal__eiij7b2h get_sprinkles_justifyContent_center__eiij7b39 get_sprinkles_whiteSpace_nowrap__eiij7b8j styles_getButtonStyles_size_square_md__1yqzead10 styles_variantSize_square_md__1yqzead5 styles__1yqzead0 get_sprinkles_fontSize_body_md__eiij7b21 get_sprinkles_lineHeight_body_md__eiij7b3f get_sprinkles_aspectRatio_square__eiij7b6 get_sprinkles_flexShrink_0__eiij7b1w get_sprinkles_height_spacing_5__eiij7b33 get_sprinkles_width_spacing_5_mobile__eiij7biw get_sprinkles_marginLeft_auto__eiij7b4d"
            name="close"
            type="button"
          >
            <div
              class="get_sprinkles_alignItems_center__eiij7b2 get_sprinkles_display_flex_mobile__eiij7b8q get_sprinkles_gap_spacing_1_mobile__eiij7ba2 get_sprinkles_color_inherit__eiij7b1p"
            >
              <div
                class="styles_getSlotStyles__hlxmy27 styles__hlxmy26 get_sprinkles_alignItems_center__eiij7b2 get_sprinkles_color_text_low_contrast__eiij7b1k get_sprinkles_display_flex_mobile__eiij7b8q get_sprinkles_justifyContent_center__eiij7b39 get_sprinkles_width_spacing_3_mobile__eiij7biq get_sprinkles_color_inherit__eiij7b1p get_sprinkles_flexShrink_0__eiij7b1w"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-xmark styles_icon__1ej8lf00 get_sprinkles_color_currentColor__eiij7b1o get_sprinkles_flexShrink_0__eiij7b1w"
                  data-icon="xmark"
                  data-prefix="fal"
                  focusable="false"
                  role="img"
                  viewBox="0 0 384 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M324.5 411.1c6.2 6.2 16.4 6.2 22.6 0s6.2-16.4 0-22.6L214.6 256 347.1 123.5c6.2-6.2 6.2-16.4 0-22.6s-16.4-6.2-22.6 0L192 233.4 59.5 100.9c-6.2-6.2-16.4-6.2-22.6 0s-6.2 16.4 0 22.6L169.4 256 36.9 388.5c-6.2 6.2-6.2 16.4 0 22.6s16.4 6.2 22.6 0L192 278.6 324.5 411.1z"
                    fill="currentColor"
                  />
                </svg>
              </div>
            </div>
          </button>
        </div>
        <div
          class="styles_dialogContentOuterStyle__l8uldv1 styles__l8uldv0 get_sprinkles_flexGrow_1__eiij7b1v get_sprinkles_overflowY_auto__eiij7b87"
        >
          <div
            class="styles_dialogContentInnerStyle__l8uldv3 styles__l8uldv2 get_sprinkles_padding_spacing_2_mobile__eiij7bc8"
          >
            hello
          </div>
        </div>
      </div>
    </dialog>
  </div>
</div>
`;
